ReportingAPI: PermissionPolicyViolationReport (#43783)
mdn/content: files/en-us/web/api/permissionspolicyviolationreport/index.md

  }
}
```

### Sending a Permissions Policy violation report to a reporting endpoint

Here we define the define reporting endpoint named `"default"` using the {{httpheader("Reporting-Endpoints")}} response header, and set the `Permissions-Policy` header to block use of the `geolocation` feature.

```http
Reporting-Endpoints: default="https://example.com/reports"
Permissions-Policy: geolocation=()
```

As before, a violation is triggered by attempting to use a blocked feature:

```js
// This should generate a Permissions Policy violation
navigator.geolocation.getCurrentPosition(
  () => {},
  () => {},
);
```

The violation report will then be sent to the default endpoint as a JSON array.
Note that the `type` is `"permissions-policy-violation"` and the `body` property is a serialization of the `PermissionsPolicyViolationReport` object.

```json
[
  {
    "age": 48512,
    "body": {
      "columnNumber": 29,
      "disposition": "enforce",
      "lineNumber": 44,
      "message": "Permissions policy violation: geolocation access has been blocked because of a permissions policy applied to the current document.",

>       "policyId": "geolocation" /* Note: Chrome server-report version of featureId*/,

Reviewer: ```suggestion
      "policyId": "geolocation", // Note: Chrome server-report version of featureId
```

    - `featureId`
      - : A string representing the [Permissions Policy directive](/en-US/docs/Web/HTTP/Reference/Headers/Permissions-Policy#directives) that was violated (for example, `"geolocation"`).
    - `lineNumber`
      - : The line number in the script at which the violation occurred, or `null` if not known.
    - `message`
      - : A string containing a human-readable description of the violation.
    - `sourceFile`
      - : A string representing the URL of the script in which the violation occurred, or `null` if not known.
        Both `columnNumber` and `lineNumber` should have non-null values if this property is not `null`.

- `type`
  - : The string `"permissions-policy-violation"`, indicating that this is a Permissions Policy violation report.
- `url`
  - : A string representing the URL of the document that generated the report.

## Description

Permissions Policy violations are reported when a document attempts to use a browser feature that is blocked by its [Permissions Policy](/en-US/docs/Web/HTTP/Guides/Permissions_Policy).

> The policy is set using the {{httpheader("Permissions-Policy")}} HTTP header, or a `<meta http-equiv="Permissions-Policy">` element.

Reviewer: ```suggestion
The policy is set using the {{httpheader("Permissions-Policy")}} HTTP header, or a `<meta http-equiv="permissions-policy">` element.
```

I would suggest using lowercase value, more common in HTML and [in the spec](https://html.spec.whatwg.org/multipage/semantics.html#pragma-directives).

You can monitor for Permissions-Policy violation reports within the page that sets the policy using the [Reporting API](/en-US/docs/Web/API/Reporting_API).
To do this you create a {{domxref("ReportingObserver")}} object to listen for reports, passing a callback method and an (optional) `options` property specifying the types of reports that you want to report on.
The callback method is then called with reports of the requested types, passing a report object.

> For Permissions Policy violations, the object will be a `PermissionsPolicyViolationReport` instance (which has the [`type`](#type) property set to `"permissions-policy-violation"`).

Reviewer: Personally, I find such sentences hard to read. Comparing these two:

> 1. `checkbox` value of the `type` attribute of the `<input>` element
> 2. `<input type="checkbox">`

…I would say inline code (object, in this case) would be easier to read.

Author: @pepelsbey You mean this? I think it is better given we have talked about the type a little above and linking is a bit pointless. Accepting this comment, but let me know if I missed your point.

```suggestion
For `Permissions-Policy` violations, the object will be a `PermissionsPolicyViolationReport` instance with `PermissionsPolicyViolationReport.type == "permissions-policy-violation"`.
```

Note that we can see the URL of the page that had its policy violated (`url`), and from `body.featureId` we can see which feature was blocked.
The `body.disposition` field shows that the violation was enforced.

```json
{
  "type": "permissions-policy-violation",
  "url": "https://example.com/",
  "body": {
    "sourceFile": "https://example.com/",
    "lineNumber": 44,
    "columnNumber": 29,
    "featureId": "geolocation",
    "disposition": "enforce",
    "message": "Permissions policy violation: geolocation access has been blocked because of a permissions policy applied to the current document."
  }
}
```

Violation reports may also be sent as a JSON object in a {{httpmethod("POST")}} request to the [reporting server endpoint](/en-US/docs/Web/API/Reporting_API#reporting_server_endpoints) named `"default"`, if one is defined.
The reporting server endpoint and its mapping to a particular URL are set using the {{httpheader("Reporting-Endpoints")}} header.

The structure of the server report is almost exactly the same as `PermissionsPolicyViolationReport`, except that it additionally includes `age` and `user_agent` fields.

> [!NOTE]

> > Chrome's server-side serialization uses `policyId` rather than `featureId` for the feature name in the body of a server report.

Reviewer: ```suggestion
> Chrome's server-side serialization uses `policyId` rather than `featureId` for the feature name in the body of a server report.
> The {{domxref("PermissionsPolicyViolationReport")}} returned by a [`ReportingObserver`](/en-US/docs/Web/API/ReportingObserver) follows the specification.
```

I would say more detailed note gives more value.

Author: Agree. I have taken this suggestion and also your other suggestion about having this below the syntax. Fixed in https://github.com/mdn/content/pull/43783/changes/accb42f28d7e200d9a627ad1b339127c237170c1

```json
[
  {
    "age": 48512,
    "body": {
      "columnNumber": 29,
      "disposition": "enforce",
      "lineNumber": 44,
      "message": "Permissions policy violation: geolocation access has been blocked because of a permissions policy applied to the current document.",
      "policyId": "geolocation",
      "sourceFile": "https://example.com/"
    },
    "type": "permissions-policy-violation",
    "url": "https://example.com/",
    "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36"
  }
]
```

## Examples

### Using the `ReportingObserver` interface

You can obtain a `PermissionsPolicyViolationReport` object by configuring your page to block a browser feature, and then attempting to use it.

In this example, we will block the Geolocation API for the current document using either the {{httpheader("Permissions-Policy")}} HTTP header:

```http
Permissions-Policy: geolocation=()
```

Or equivalently via an HTML `<meta>` element:

```html

> <meta http-equiv="Permissions-Policy" content="geolocation=()" />

Reviewer: ```suggestion
<meta http-equiv="permissions-policy" content="geolocation=()" />
```

  "body": {
    "sourceFile": "https://example.com/",
    "lineNumber": 44,
    "columnNumber": 29,
    "featureId": "geolocation",
    "disposition": "enforce",
    "message": "Permissions policy violation: geolocation access has been blocked because of a permissions policy applied to the current document."
  }
}
```

### Sending a Permissions Policy violation report to a reporting endpoint

Here we define the define reporting endpoint named `"default"` using the {{httpheader("Reporting-Endpoints")}} response header, and set the `Permissions-Policy` header to block use of the `geolocation` feature.

```http
Reporting-Endpoints: default="https://example.com/reports"
Permissions-Policy: geolocation=()
```

As before, a violation is triggered by attempting to use a blocked feature:

```js
// This should generate a Permissions Policy violation
navigator.geolocation.getCurrentPosition(
  () => {},
  () => {},
);
```

The violation report will then be sent to the default endpoint as a JSON array.
Note that the `type` is `"permissions-policy-violation"` and the `body` property is a serialization of the `PermissionsPolicyViolationReport` object.

> [!NOTE]

> > Chrome's server-side serialization uses `policyId` rather than `featureId` for the feature name.

Reviewer: It would be great to mention how big of a problem this incompatibility is and even suggest what the reader could do about it. And maybe it would be better to mention it just once but in more detail right after listing the instance properties.

Author: Fixed in https://github.com/mdn/content/pull/43783/changes/accb42f28d7e200d9a627ad1b339127c237170c1
